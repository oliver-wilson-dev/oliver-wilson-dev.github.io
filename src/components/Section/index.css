@import "../../styles/transitions.css";

@keyframes fadeIn {
    0% { 
        opacity: 0; 
    }
    100% { 
        opacity: 1; 
    }
}

.section {
    transition: visibility var(--sectionVisibilityTransitionDuration), opacity var(--sectionOpacityTransitionDuration);
    transition-delay: var(--spinnerOpacityTransitionDuration);
    opacity: 0;
    visibility: hidden;
    &.show {
        opacity: 1;
        visibility: visible;
      }
}

.displayBox {
    background: var(--primaryBackground);
    -webkit-box-shadow: 8px 8px 0px var(--primaryBorderColour);
    -moz-box-shadow: 8px 8px 0px var(--primaryBorderColour);
    box-shadow: 8px 8px 0px var(--primaryBorderColour);
    border: 3px solid var(--primaryBorderColour);
    padding: 10px;
    max-height: 250px;
    overflow: hidden;
    position: relative;
}

.displayBoxExpanded {
    max-height: 1500px;
}

.collapseExpand {
    margin: 5px 5px 5px auto;
    position: absolute;
    right: 0px;
    top: 0px;
    font-size: 30px;
    width: 35px;
    border: none;
    background: transparent;
    font-family: inherit;
    font-weight: 900;
    color: inherit;
}

@media all and (min-width: 786px) {
    .landingBlurb {
        max-width: 800px;
    }
}