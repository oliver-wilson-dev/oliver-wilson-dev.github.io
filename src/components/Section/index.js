import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import styles from './index.css';
import sharedStyles from '../App/index.css';

class Section extends React.Component {
  constructor(props) {
    super(props);

    this.displayBoxRef = null;
    this.childrenRef = React.createRef();

    this.state = {
      expanded: false,
      expandable: false,
      visible: false
    };

    this.setDisplayBoxRef = this.setDisplayBoxRef.bind(this);

    this.canBeFocusedProgrammatically = -1;
  }


  componentDidMount() {
    this.setState(
      previousState => ({
        ...previousState,
        expandable: this.displayBoxRef.offsetHeight >= 250,
        visible: true
      })
    );
  }

  componentDidUpdate() {
    const { state: { expanded } } = this;

    if (expanded) { this.childrenRef.current.focus(); }
  }

  setDisplayBoxRef(elm) {
    this.displayBoxRef = elm;
  }

  onButtonClick = (e) => {
    e.preventDefault();
    this.setState(previousState => ({
      ...previousState,
      expanded: !previousState.expanded
    }));
  }

  render() {
    const {
      canBeFocusedProgrammatically,
      onButtonClick,
      setDisplayBoxRef,
      props: { title, children },
      state: { expanded, expandable, visible },
    } = this;

    return (
      <div className={cn(sharedStyles.flexColumn, sharedStyles.flexCenter, styles.section, {
        [styles.show]: visible
      })}
      >
        <h2 className={sharedStyles.sectionTitle}>{title}</h2>
        <div
          ref={setDisplayBoxRef}
          className={(cn(styles.landingBlurb, styles.displayBox,
            {
              [styles.displayBoxExpanded]: expanded,
              [styles.blurBottom]: expandable && !expanded
            }))
          }
        >
          {expandable && (
          <button
            className={styles.collapseExpand}
            type="button"
            onClick={onButtonClick}
            aria-expanded={expanded}
            aria-label={expanded ? 'click to minimise this section' : 'click to expand this section to reveal more content'}
          >
            <span>{expanded ? 'âˆ’' : '+'}</span>
          </button>
          )}
          <div
            ref={this.childrenRef}
            tabIndex={canBeFocusedProgrammatically}
            aria-hidden={expandable && !expanded}
          >
            {children}
          </div>
        </div>
      </div>
    );
  }
}

Section.propTypes = {
  title: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired
};

export default Section;
