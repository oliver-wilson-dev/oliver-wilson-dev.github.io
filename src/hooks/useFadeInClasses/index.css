@import "../../styles/transitions.css";

.hiddenFadeIn {
    transition: visibility var(--sectionVisibilityTransitionDuration), opacity var(--sectionOpacityTransitionDuration);
    transition-delay: var(--spinnerOpacityTransitionDuration);
    opacity: 0;
    visibility: hidden;

    &.show {
        opacity: 1;
        visibility: visible;
    }
}