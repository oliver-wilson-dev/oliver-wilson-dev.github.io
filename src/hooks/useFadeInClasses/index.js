import { useEffect, useState } from 'react';
import sectionStyles from './index.css';

const useFadeInClasses = () => {
  const [visible, setVisible] = useState(false);

  useEffect(() => {
    setVisible(true);
  }, []);

  const fadeInClasses = {
    [sectionStyles.hiddenFadeIn]: true,
    [sectionStyles.show]: visible
  };

  return { fadeInClasses };
};

export default useFadeInClasses;
