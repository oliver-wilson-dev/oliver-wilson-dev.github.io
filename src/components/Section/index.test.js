import React from 'react';
import { mount } from 'enzyme';
import Section from '.';
import useFadeInClasses from '../../hooks/useFadeInClasses';

jest.mock('../SectionSlider', () => {
  const SectionTile = ({ children }) => <div>{children}</div>;

  return SectionTile;
});

jest.mock('../SectionTile', () => {
  const SectionTile = () => null;

  return SectionTile;
});

jest.mock('../../hooks/useFadeInClasses');

const defaultProps = {
  title: 'test-title',
  children: [<p key="test-key">this is some content</p>]
};

const render = ({ overrideProps } = { overrideProps: {} }) => mount(
  <Section
    {...defaultProps}
    {...overrideProps}
  />
);

describe('<Section/>', () => {
  beforeEach(() => {
    useFadeInClasses.mockReturnValue({ fadeInClasses: 'test-fade-in-classes' });
  });


  describe('when the children are HTML elements', () => {
    it('should exist', () => {
      expect(render().exists()).toBe(true);
    });

    it('should render correctly', () => {
      expect(render()).toMatchSnapshot();
    });
  });

  describe('when the children are react components', () => {
    const SomeComponent = () => <p>some component</p>;
    it('should exist', () => {
      expect(render({
        overrideProps: {
          children: [<SomeComponent key="1" />, <SomeComponent key="2" />]
        }
      }).exists()).toBe(true);
    });

    it('should render correctly', () => {
      expect(render({
        overrideProps: {
          children: [<SomeComponent key="1" />, <SomeComponent key="2" />]
        }
      })).toMatchSnapshot();
    });
  });

  describe('when the children are both components and HTML elements', () => {
    const htmlElement = <p key="test-key">some component</p>;
    const SomeComponent = () => htmlElement;

    it('should exist', () => {
      expect(render({
        overrideProps: {
          children: [<SomeComponent />, htmlElement]
        }
      }).exists()).toBe(true);
    });

    it('should render correctly', () => {
      expect(render({
        overrideProps: {
          children: [<SomeComponent key="1" />, htmlElement]
        }
      })).toMatchSnapshot();
    });
  });
});
