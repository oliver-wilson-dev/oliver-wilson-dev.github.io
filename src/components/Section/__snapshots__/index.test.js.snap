// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Section/> should render correctly 1`] = `
<Section
  title="test-title"
>
  <div
    className="flexColumn flexCenter section show"
  >
    <h2
      className="sectionTitle"
    >
      test-title
    </h2>
    <div
      className="landingBlurb displayBox"
    >
      <p>
        this is some content
      </p>
    </div>
  </div>
</Section>
`;

exports[`<Section/> when the display box content should be expandable and the section is expanded should render correctly 1`] = `
<Section
  title="test-title"
>
  <div
    className="flexColumn flexCenter section show"
  >
    <h2
      className="sectionTitle"
    >
      test-title
    </h2>
    <div
      className="landingBlurb displayBox displayBoxExpanded"
    >
      <button
        className="collapseExpand"
        onClick={[Function]}
        type="button"
      >
        <span>
          âˆ’
        </span>
      </button>
      <p>
        this is some content
      </p>
    </div>
  </div>
</Section>
`;

exports[`<Section/> when the display box content should be expandable should render an expandable button 1`] = `
<Section
  title="test-title"
>
  <div
    className="flexColumn flexCenter section show"
  >
    <h2
      className="sectionTitle"
    >
      test-title
    </h2>
    <div
      className="landingBlurb displayBox"
    >
      <button
        className="collapseExpand"
        onClick={[Function]}
        type="button"
      >
        <span>
          +
        </span>
      </button>
      <p>
        this is some content
      </p>
    </div>
  </div>
</Section>
`;
